### PR TITLE
Add contact app [WIP]

### PR DESCRIPTION
**Pull request by daniel-clayton**
Needs testing.

### CODE REVIEW

File: {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/apps/contact/models.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
+from cms.apps.pages.models import ContentBase
+from cms.models import HtmlField
+from django.db import models
+
+
+class Contact(ContentBase):
+
```
Comment (lewiscollard) [inline {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/apps/contact/models.py]: You'll probably want to add a content type icon here. See https://github.com/onespacemedia/cms-icons/

File: {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/apps/contact/admin.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        response['Content-Type'] = 'text/csv'
+        response['Content-Disposition'] = (
+            'attachment;filename="components-bureau-submissions-{}.csv'.format(
+                now().isoformat()
+            )
+        )
+        return response
+
+    def get_urls(self):
+        patterns = super(ContactSubmissionAdmin, self).get_urls()
```
Comment (lewiscollard) [inline {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/apps/contact/admin.py]: You should prefer the simpler Python 3 `super()` form here.

File: {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/apps/contact/models.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    icon = 'cms-icons/form.png'
+
+    call_to_action = models.ForeignKey(
+        'components.CallToAction',
+        blank=True,
+        null=True,
+        on_delete=models.PROTECT,
+    )
+
+    hero_kicker = models.CharField(
```
Comment (lewiscollard) [inline {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/apps/contact/models.py:25]: As these are rendered in a "Hero" fieldset, you should give this a `verbose_name` of 'kicker'.

File: {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/apps/contact/models.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        blank=True,
+        null=True,
+        help_text='If this is left blank it will use the page title',
+    )
+
+    hero_title = models.CharField(
+        max_length=255,
+    )
+
+    form_email_address = models.CharField(
```
Comment (lewiscollard) [inline {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/apps/contact/models.py]: Same applies for these form_* fields here as applied to the hero - as they're contained in a fieldset called "Form" the default rendering of "Form email address" is redundant.

Comment (lewiscollard) [reply {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/apps/contact/models.py]: Also, unless you are planning on making this accept multiple email addresses separated by something, you should make this an EmailField.

On the other hand, if you are planning on that, put a method on the content type to do this splitting.

File: {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/apps/contact/models.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        help_text='This is the email address form submissions will be sent to.',
+    )
+
+    form_title = models.CharField(
+        max_length=150,
+        blank=True,
+        null=True,
+    )
+
+    success_page_title = models.CharField(
```
Comment (lewiscollard) [inline {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/apps/contact/models.py]: Same comments apply re: fieldset titles above - give this a simpler `verbose_name`.

File: {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/apps/contact/models.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        blank=True,
+        null=True,
+        on_delete=models.PROTECT,
+    )
+
+    hero_kicker = models.CharField(
+        max_length=100,
+        blank=True,
+        null=True,
+        help_text='If this is left blank it will use the page title',
```
Comment (lewiscollard) [inline {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/apps/contact/models.py]: This `help_text` does not match the behaviour of the template.

Comment (lewiscollard) [reply {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/apps/contact/models.py]: Also, punctuate `help_text`s as sentences.

File: {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/apps/contact/models.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+    )
+
+    hero_title = models.CharField(
+        max_length=255,
+    )
+
+    form_email_address = models.CharField(
+        max_length=255,
+        help_text='This is the email address form submissions will be sent to.',
```
Comment (lewiscollard) [inline {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/apps/contact/models.py]: Minor grammatical nitpick, but I'd prefer the simpler wording of "Form submissions will be sent to this email address."

**Review by lewiscollard — CHANGES_REQUESTED** [review]

File: {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/apps/contact/models.py [hunk 8]
```diff
@@ -74,6 +74,20 @@ class Contact(ContentBase):
     def __str__(self):
         return self.page.title
 
+    @property
+    def get_email_addresses(self):
```
Comment (lewiscollard) [inline {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/apps/contact/models.py]: Properties should normally be named as if they are instance attributes. Either name this `email_addresses` or keep the existing name and make it a normal function.

File: {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/apps/contact/admin.py [hunk 9]
```diff
@@ -16,6 +18,11 @@
 class ContactSubmissionAdmin(VersionAdmin):
     list_display = ['first_name', 'last_name', 'email', 'phone_number', 'job_title', 'reason_for_enquiry', 'created']
 
+    def clean(self):
+        for email in self.get_email_addresses:
```
Comment (lewiscollard) [inline {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/apps/contact/admin.py]: `self.get_email_addresses` is not a property or attribute of `ContactSubmissionAdmin`.

**Review by lewiscollard — CHANGES_REQUESTED** [review]

File: {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/apps/contact/models.py [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+        return super().clean(self)
+
+    @property
+    def email_addresses(self):
+        # Strip out commas if comma separated.
+        emails = self.form_email_address.replace(',', ' ')
+
+        emails = emails.split(' ')
+        for email in emails:
+            email.strip()
```
Comment (lewiscollard) [inline {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/apps/contact/models.py]: This does nothing; `.strip()` does not modify in place. A unit test should have caught this.

Comment (lewiscollard) [reply {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/apps/contact/models.py]: Easier: `re.split('[\s,]+', self.form_email_address)`

File: {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/apps/contact/models.py [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+
+    hero_title = models.CharField(
+        max_length=255,
+        verbose_name='Title',
+    )
+
+    form_email_address = models.CharField(
+        max_length=255,
+        help_text='Form submissions will be sent to these addresses. Separate multiple emails with a comma or space',
+        verbose_name='Email Address',
```
Comment (lewiscollard) [inline {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/apps/contact/models.py]: We never use Title Case.

**Comment by lewiscollard** [comment]
I am setting a bad example, but I've rebased and force-pushed to clean up the history of this one. Hopefully I did not murder anything.

**Review by lewiscollard — APPROVED** [review]

**Comment by AidanCurrah** [comment]
Since we're adding the landing hero macro in this PR, should we change the other pages to use the macro-ized hero?

**Pull request merged**